### PR TITLE
Make sure Byron header is entirely strict

### PR DESCRIPTION
**Pull request by edsko**
The presence of an `Either` was causing a memory leak.

### CODE REVIEW

File: ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Byron.hs [hunk 1]
```diff
@@ -466,34 +464,27 @@ mkByronHeaderOrEBB :: Either (CC.Block.ABoundaryHeader ByteString)
...
-      -- Since the thunk to compute the hash (see 'mkByronHeaderOrEBB') only
-      -- refers to 'byronHeaderOrEBB', this thunk should not secretly hold on
-      -- to more data than 'byronHeaderOrEBB' itself holds on to.
-    } deriving (Eq, Show)
+  --
+  -- This guarantees that the lazily evaluated hash won't retain anything more
+  -- than is retained /anyway/ by the header itself.
+  data Header (ByronBlockOrEBB cfg) =
+        ByronHeaderRegular  !(CC.Block.AHeader         ByteString) ByronHash
+      | ByronHeaderBoundary !(CC.Block.ABoundaryHeader ByteString) ByronHash
```
Comment (intricate) [inline ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Byron.hs]: :+1:

Comment (avieth) [reply ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Byron.hs]: Are these constructors exported?

Comment (avieth) [reply ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Byron.hs]: #1053

**Pull request merged**